### ISSUE TITLE
Create Primary/Foreign Keys

### ISSUE THREAD

**Issue by zzeekk** [smart-data-lake/smart-data-lake#865]
**Is your feature request related to a problem? Please describe.**
SDLB allows to define Primary/Foreign Keys for TableDataObjects, e.g. DeltaLakeTableDataObject.
Primary Key is used by Deduplicate/HistorizeAction. Foreign Keys are just an information for now.
I would like SDLB to create PKs/FKs on the underlying catalog of DataObjects, e.g. Jdbc Database, Unity Catalog or Snowflake Catalog.

**Describe the solution you'd like**
The following TableDataObjects should execute DDL statements to create PKs/FKs:
- JdbcTableDataObject
- SnowflakeTableDataObject
- DeltaLakeTableDataObject if UCFileSystemFactory.isDatabricksEnv = true

If the DeltaLakeTableDataObject has a UniqueKeyExpectation for the Primary Key (e.g. key=empty), then create the PK with additional RELY keyword for query optimization, see also https://docs.databricks.com/en/sql/user/queries/query-optimization-constraints.html

To be discussed when PKs/FKs should be created and synchronized/updated if there was changed in configuration, e.g. additional FK.
Should we synchronize this on every write, or have a separate command line application which can be started after a release?

**Comment by zzeekk** [comment]
In addition it would be nice to update column/table columns within the same process.

For Databricks the SQL commands are as follows:

Primary and Foreign Keys
```
ALTER TABLE dev.test.delta_comment_test ALTER COLUMN num SET NOT NULL;
ALTER TABLE dev.test.delta_comment_test ADD CONSTRAINT pk PRIMARY KEY(num) RELY;
ALTER TABLE dev.test.delta_comment_test ADD CONSTRAINT fk1 FOREIGN KEY(num) REFERENCES dev.test.delta_comment_test;
SELECT *  FROM dev.information_schema.table_constraints
WHERE constraint_type='FOREIGN KEY' and table_catalog='dev' and table_schema='test' and table_name='delta_comment_test';
SELECT *  FROM dev.information_schema.table_constraints
WHERE constraint_type='PRIMARY KEY' and table_catalog='dev' and table_schema='test' and table_name='delta_comment_test';
```

Comments
```
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types.MetadataBuilder
val df = Seq(("a",1)).toDF("str","num")
  .withColumn("num", col("num").as("num", new MetadataBuilder().putString("comment","number").build()))
df.write.saveAsTable("dev.test.delta_comment_test")
ALTER TABLE dev.test.delta_comment_test SET TBLPROPERTIES ('comment' = 'A table comment.');
```

To be tested if column comments are updated if changed? Check also if this works with every SaveMode (e.g. Append) or only when creating a new table?

**Comment by iliassoto** [comment]
Creation and Update of PKs for JdbcTableDataObject should be done in the prepare phase, for SnowflakeTableDataObject and DeltaLakeTableDataObject in the Exec phase.
For Foreign Keys, the creation and update should always be done in the exec phase (to avoid having referenced tables that not exist yet).

**Comment by iliassoto** [comment]
Comments in DeltaLake are also supported directly using SQL, see https://docs.delta.io/latest/delta-batch.html#update-table-schema

**Comment by iliassoto** [comment]
As discussed, this issue and respective PR will only implement PKs and comments. Updates of FKs and their behaviour are to be discussed in a separate issue.

**Comment by iliassoto** [comment]
JdbcTableDataObject will not implement the comment function, as this is not a part of the INFORMATION_SCHEMA standard. Therefore, depending on the database / driver used, it is better to define comments as a postExecSql statement.

**Comment by zzeekk** [comment]
> Comments in DeltaLake are also supported directly using SQL, see https://docs.delta.io/latest/delta-batch.html#update-table-schema

True, but it is also nice to define comments in the DataFrame of a Transformation as shown below, or as Dataset[ProductXyz] with ScalaDoc comments on the case class (not working yet).

> val df = Seq(("a",1)).toDF("str","num")
>   .withColumn("num", col("num").as("num", new MetadataBuilder().putString("comment","number").build()))
> df.write.saveAsTable("dev.test.delta_comment_test")

Did you test if columns are updated in the DB metadata when changed?
It would also be interesting to check if this works for JDBCTableDataObject. I know that it works for HiveTableDataObjects.

Drawback is that you can not add columns for SQL transformations like that... but maybe we could also use the schema to manage columns, e.g. [/* this is a comment */](https://docs.oracle.com/en/database/other-databases/nosql-database/24.1/java-driver-table/ddl-comments.html). To be tested if Spark parses this into the StructType...
